packages/core/src/codewhispererChat/tools/fsRead.ts

```diff
@@ -0,0 +1,175 @@
+                throw new Error(`Path: "${this.fsPath}" does not exist or cannot be accessed. (${err})`)
+            }
+
+            if (fileStat.type === vscode.FileType.File) {
+                const fileContents = await this.readFileFully(fileUri)
+                getLogger().info(`Read file: ${this.fsPath}, size: ${fileContents.length}`)
+                return this.handleFileRange(fileContents)
+            } else if (fileStat.type === vscode.FileType.Directory) {
+                const maxDepth = this.getDirectoryDepth() ?? 0
+                const listing = await this.readDirectoryRecursive(fileUri, maxDepth)
+                return this.makeOutput(listing.join('\n'))
+            } else {
+                throw new Error(`"${this.fsPath}" is neither a standard file nor directory.`)
+            }
+        } catch (error: any) {
+            getLogger().error(`[fs_read] Failed to read "${this.fsPath}": ${error.message || error}`)
+            throw new Error(`[fs_read] Failed to read "${this.fsPath}": ${error.message || error}`)
+        }
+    }
+
+    private async readFileFully(fileUri: vscode.Uri): Promise<string> {
+        getLogger().info(`Reading file: ${fileUri.fsPath}`)
+        const data = await vscode.workspace.fs.readFile(fileUri)
+        return Buffer.from(data).toString('utf8')
+    }
+
+    private handleFileRange(fullText: string): InvokeOutput {
+        if (!this.readRange || this.readRange.length === 0) {
+            getLogger().info('No range provided. returning entire file.')
+            return this.makeOutput(this.enforceMaxSize(fullText))
+        }
+
+        const lines = fullText.split('\n')
+        const [start, end] = this.parseLineRange(lines.length, this.readRange)
+        if (start > end) {
+            getLogger().error(`Invalid range: ${this.readRange.join('-')}`)
+            return this.makeOutput('')
+        }
+
+        getLogger().info(`Reading file: ${this.fsPath}, lines ${start + 1}-${end + 1}`)
+        const slice = lines.slice(start, end + 1).join('\n')
+        return this.makeOutput(this.enforceMaxSize(slice))
+    }
+
+    private parseLineRange(lineCount: number, range: number[]): [number, number] {
+        const startIdx = range[0]
+        let endIdx = range.length >= 2 ? range[1] : undefined
+
+        if (endIdx === undefined) {
+            endIdx = -1
+        }
+
+        const convert = (i: number): number => {
+            return i < 0 ? lineCount + i : i - 1
+        }
+
+        const finalStart = Math.max(0, Math.min(lineCount - 1, convert(startIdx)))
+        const finalEnd = Math.max(0, Math.min(lineCount - 1, convert(endIdx)))
+        return [finalStart, finalEnd]
+    }
+
+    private getDirectoryDepth(): number | undefined {
+        if (!this.readRange || this.readRange.length === 0) {
+            return 0
+        }
+        return this.readRange[0]
+    }
+
+    private async readDirectoryRecursive(dirUri: vscode.Uri, maxDepth: number): Promise<string[]> {
```

Comment:
this might be useful to have in our workspaceUtils.ts if this doesn't exist already

Reply:
Moved to utility.